selfservice/strategy/code/strategy_recovery_test.go

```diff
@@ -0,0 +1,1003 @@
+	})
+
+	t.Run("description=should fail on negative expiry time", func(t *testing.T) {
+		_, _, err := createCode(x.NewUUID().String(), pointerx.String("-1h"))
+
+		require.IsType(t, err, new(kratos.GenericOpenAPIError), "%T", err)
+		assert.EqualError(t, err.(*kratos.GenericOpenAPIError), "400 Bad Request")
+
+		reason := *err.(*kratos.GenericOpenAPIError).Model().(client.JsonError).Error.Reason
+
+		assert.Contains(t, reason, `Value from "expires_in" must result to a future time:`)
+	})
+
+	t.Run("description=should create code without email", func(t *testing.T) {
+		id := identity.Identity{Traits: identity.Traits(`{}`)}
+
+		require.NoError(t, reg.IdentityManager().Create(context.Background(),
+			&id, identity.ManagerAllowWriteProtectedTraits))
+
+		code, _, err := createCode(id.ID.String(), nil)
+		require.NoError(t, err)
+
+		require.NotEmpty(t, code.RecoveryLink)
+		require.Contains(t, code.RecoveryLink, "flow=")
+		require.NotContains(t, code.RecoveryLink, "code=")
+		require.NotEmpty(t, code.RecoveryCode)
+		require.True(t, code.ExpiresAt.Before(time.Now().Add(conf.SelfServiceFlowRecoveryRequestLifespan(ctx))))
+
+		res, err := publicTS.Client().Get(code.RecoveryLink)
+		require.NoError(t, err)
+		body := ioutilx.MustReadAll(res.Body)
+
+		action := gjson.GetBytes(body, "ui.action").String()
+		require.NotEmpty(t, action)
+		csrfToken := gjson.GetBytes(body, "ui.nodes.#(attributes.name==csrf_token).attributes.value").String()
+		require.NotEmpty(t, csrfToken)
+
+		res, err = publicTS.Client().PostForm(action, url.Values{
+			"csrf_token": {csrfToken},
+			"code":       {code.RecoveryCode},
+		})
+		body = ioutilx.MustReadAll(res.Body)
+
+		// Recovery successful - redirected to Settings
+		assert.Equal(t, text.NewRecoverySuccessful(time.Now().Add(time.Hour)).Text,
+			gjson.GetBytes(body, "ui.messages.0.text").String())
+	})
+
+	t.Run("description=should not be able to recover with expired code", func(t *testing.T) {
+		recoveryEmail := "recover.expired@ory.sh"
+		id := identity.Identity{Traits: identity.Traits(fmt.Sprintf(`{"email":"%s"}`, recoveryEmail))}
+
+		require.NoError(t, reg.IdentityManager().Create(context.Background(),
+			&id, identity.ManagerAllowWriteProtectedTraits))
+
+		code, _, err := createCode(id.ID.String(), pointerx.String("100ms"))
+		require.NoError(t, err)
+
+		time.Sleep(time.Millisecond * 100)
+		require.NotEmpty(t, code.RecoveryLink)
+		require.True(t, code.ExpiresAt.Before(time.Now().Add(conf.SelfServiceFlowRecoveryRequestLifespan(ctx))))
+
+		res, err := publicTS.Client().Get(code.RecoveryLink)
+		body := ioutilx.MustReadAll(res.Body)
+
+		action := gjson.GetBytes(body, "ui.action").String()
+		require.NotEmpty(t, action)
+		csrfToken := gjson.GetBytes(body, "ui.nodes.#(attributes.name==csrf_token).attributes.value").String()
+		require.NotEmpty(t, csrfToken)
+
+		res, err = publicTS.Client().PostForm(action, url.Values{
```

Comment:
Make this a helper and DRY it up in the tests

Reply:
Maybe explore testhelpers.Submit* helpers

Reply:
I DRYed up the tests as best, as I could without refactoring everything :D please have another look.